mri/reconstruct/fourier.py

```diff
@@ -521,3 +521,114 @@ def adj_op(self, coeffs):
+            must be carried out. Please refer to Documentation of
+            NoncartesianFFT
+        """
+        self.num_slices = shape[2]
+        (plane_samples, self.z_samples, self.stack_len,
+         self.acq_num_slices, self.sort_pos) = get_stacks_fourier(kspace_loc)
+        self.FT = NonCartesianFFT(samples=plane_samples, shape=shape[0:2],
+                                  implementation=implementation)
+        self.nb_coils = n_coils
+
+    def _op(self, data):
+        first_fft = np.fft.fftshift(np.fft.fft(np.fft.ifftshift(data, axes=2),
+                                               n=self.acq_num_slices,
+                                               norm="ortho"),
+                                    axes=2)
+        stacked_kspace = np.asarray(
+            [self.FT.op(first_fft[:, :, slice])
+             for slice in np.arange(self.acq_num_slices)])
+        stacked_kspace = np.reshape(stacked_kspace,
+                                    self.acq_num_slices * self.stack_len)
+        # Unsort the Coefficients
+        inv_idx = np.zeros_like(self.sort_pos)
+        inv_idx[self.sort_pos] = np.arange(len(self.sort_pos))
+        return stacked_kspace[inv_idx]
+
+    def op(self, data):
+        """ This method calculates Fourier transform.
+        Parameters
+        ----------
+        img: np.ndarray
+            input image as array.
+
+        Returns
+        -------
+        result: np.ndarray
+            Forward 3D Fourier transform of the image.
+        """
+        if self.nb_coils == 1:
+            coeff = self._op(np.squeeze(data))
+        else:
+            coeff = [self._op(data[i])
+                     for i in range(self.nb_coils)]
+        coeff = np.asarray(coeff)
+        return coeff
+
+    def _adj_op(self, coeff):
+        coeff = coeff[self.sort_pos]
+        stacks = np.reshape(coeff, (self.acq_num_slices, self.stack_len))
+        first_fft = np.asarray([self.FT.adj_op(stacks[slice]).T
+                                for slice in np.arange(stacks.shape[0])])
+        # TODO fix for higher N, this is not a usecase
+        # interpolate_kspace = interp1d(self.z_samples[:, 0],
+        #                               first_fft, kind='zero',
+        #                               axis=0, bounds_error=False,
+        #                               fill_value=0)
+        # first_fft = interpolate_kspace(np.linspace(-0.5, 0.5,
+        #                                            self.num_slices,
+        #                                            endpoint=False))
+        final = np.fft.ifftshift(np.fft.ifft(
+            np.asarray(np.fft.fftshift(first_fft, axes=0)),
+            axis=0, n=self.num_slices, norm="ortho"),
+            axes=0)
+        return final.T
+
+    def adj_op(self, coeff):
+        """ This method calculates inverse masked non-uniform Fourier
+        transform of a 1-D coefficients array.
+        Parameters
+        ----------
+        x: np.ndarray
```

Comment:
watch out for all the parameters/docs mismatches (this one is the last one I highlight but may be more down the road)